keps/sig-network/1672-tracking-terminating-endpoints/README.md

```diff
@@ -0,0 +1,152 @@
+
+One example would be the IPVS proxier which should set the weight of an endpoint to 0
+during termination and finally remove the real server when the endpoint is removed.
+Without knowing when a pod is done terminating, the IPVS proxy makes a best-effort guess
+at when the pod is terminated by looking at the connection tracking table.
+
+### Notes/Constraints/Caveats (optional)
+
+### Risks and Mitigations
+
+Tracking the terminating state of endpoints poses some scalability concerns as each
+terminating endpoint adds additional writes to the API. Today, a terminating pod
+results in 1 write in Endpoints (removing the endpoint). With the proposed changes,
+each terminating endpoint could result in at least 2 writes (ready -> terminating -> removed)
+and possibly more depending on how many times readiness changes during termination.
+
+## Design Details
+
+To track whether an endpoint is terminating, a `terminating` field would be added as part of
+the `EndpointCondition` type in the EndpointSlice API.
+
+```go
+// EndpointConditions represents the current condition of an endpoint.
+type EndpointConditions struct {
+    // ready indicates that this endpoint is prepared to receive traffic,
+    // according to whatever system is managing the endpoint. A nil value
+    // indicates an unknown state. In most cases consumers should interpret this
+    // unknown state as ready.
+    // +optional
+    Ready *bool `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+
+    // terminating indicates if this endpoint is terminating. Consumers should assume a
+    // nil value indicates the endpoint  is not terminating.
+    // +optional
+    Terminating *bool `json:"terminating,omitempty" protobuf:"bytes,2,name=terminating"`
+}
+```
+
+NOTE: A nil value for `Terminating` indicates that the endpoint is not terminating.
+
+Updates to endpointslice controller:
+* include pods with a deletion timestamp in endpointslice
+* any pod with a deletion timestamp will have condition.terminating = true
+* allow endpoint ready condition to change during termination
+
+### Test Plan
+
+endpointslice controller unit tests:
+* Unit tests will validate pods with a deletion timestamp are included with condition.teriminating = true
+* Unit tests will validate that the ready condition can change for terminating endpoints
+
+There will be no e2e tests since consumption of this new API is out-of-scope for this KEP.
+Any future KEP that consumes this API should have e2e tests to ensure behavior for terminating
+endpoints is correct.
+
+### Graduation Criteria
+
+Since this is an addition to the EndpointSlice API, graduation will follow the graduation
+timeline for the [EndpointSlice API work](/keps/sig-network/20190603-EndpointSlice-API.md).
+
+### Upgrade / Downgrade Strategy
+
+Since this is an addition to the EndpointSlice API, the upgrade/downgrade strategy will follow that
+of the [EndpointSlice API work](/keps/sig-network/20190603-EndpointSlice-API.md).
```

Comment:
This and other links to the EndpointSlice KEP are out of date: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20190603-endpointslices/README.md.

Reply:
fixed